File: LeetCode-Cpp/MinimumWindowSubstring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int i = 0;
        int j = 0;
        int minlen = s.size() + 1; 
        unordered_map<char, int> mp;
        for (int i = 0; i < t.size(); i++) {
            mp[t[i]]++;
        }
        int cnt = mp.size();
        int start = 0;

        while (j < s.size()) {
            if (mp.find(s[j]) != mp.end()) {
                mp[s[j]]--; 
                if (mp[s[j]] == 0) {
                    cnt--;
                }
            }

            while (cnt == 0) {
                if (j - i + 1 < minlen) {
                    minlen = j - i + 1;
                    start = i;
                }

                if (mp.find(s[i]) != mp.end()) {
                    mp[s[i]]++;
                    if (mp[s[i]] > 0) {
                        cnt++;
                    }
                }
                i++;
            }

            j++;
        }

        return minlen == s.size() + 1 ? "" : s.substr(start, minlen);
    }
};

File: LeetCode-Python/MinimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""

        # Initialize the character count dictionaries for t and the current window in s
        t_count = Counter(t)
        window_count = Counter()

        # Variables to keep track of the start and end of the minimum window substring
        left = 0
        right = 0

        # Variables to keep track of the minimum window substring
        min_len = float('inf')
        min_window = ""

        # Variable to track the number of characters in t that are missing in the current window
        required_chars = len(t)

        while right < len(s):
            # Expand the window to the right
            if s[right] in t_count:
                window_count[s[right]] += 1
                if window_count[s[right]] <= t_count[s[right]]:
                    required_chars -= 1

            # Contract the window from the left
            while required_chars == 0:
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    min_window = s[left:right + 1]

                if s[left] in t_count:
                    window_count[s[left]] -= 1
                    if window_count[s[left]] < t_count[s[left]]:
                        required_chars += 1

                left += 1

            # Move the right pointer to expand the window further
            right += 1

        return min_window
